### PR TITLE
jesd204 framework wip

### PR DESCRIPTION
**Pull request by buha**
this is a draft for @dbogdan 

It currently produces the following output with LOG_LEVEL=7 (+ optionally, the modification about the ad9081 vcxo frequency):
```
Hello
rx_adxcvr: OK (10000000 kHz)
tx_adxcvr: OK (10000000 kHz)
ad9081_jesd_rx_link_status_print: JESD TX (JRX) Link1 0x0 lanes in DATA
tx_adxcvr: OK (10000000 kHz)
ad9081_jesd_rx_link_status_print: JESD TX (JRX) Link1 0xF lanes in DATA
ad9081_jesd_tx_link_status_print: JESD RX (JTX) Link1 in DATA, SYNC deasserted, PLL locked, PHASE established, MODE valid
DAC IRQ status 0xf043000040
AD9081 Rev. 1 Grade 10 (API 1.0.7) probed
op 0
link 0
link 1
op 1
link 0
DEBUG: ad9081_jesd204_link_init:271 link_num 0 reason initialization
link 1
DEBUG: ad9081_jesd204_link_init:271 link_num 0 reason initialization
op 2
link 0
DEBUG: hmc7044_jesd204_link_supported:890 link_num 0 reason initialization
DEBUG: hmc7044_lmfc_lemc_validate: dividend=3000000000 divisor=7812500 GCD=7812500 (hmc->pll2_freq=3000000000, min=732780)
DEBUG: hmc7044_jesd204_link_supported:904 link_num 0 LMFC/LEMC 7812500/7812500 gcd 7812500
link 1
DEBUG: hmc7044_jesd204_link_supported:890 link_num 0 reason initialization
DEBUG: hmc7044_lmfc_lemc_validate: dividend=7812500 divisor=7812500 GCD=7812500 (hmc->pll2_freq=3000000000, min=732780)
DEBUG: hmc7044_jesd204_link_supported:904 link_num 0 LMFC/LEMC 7812500/7812500 gcd 7812500
op 3
link 0
DEBUG: hmc7044_jesd204_link_pre_setup:1084 link_num 0 reason initialization
DEBUG: hmc7044_jesd204_link_pre_setup:1100 Found SYSREF channel2 setting f=1953125 Hz
DEBUG: hmc7044_jesd204_link_pre_setup:1100 Found SYSREF channel10 setting f=1953125 Hz
WARNING: hmc7044_jesd204_link_pre_setup: Link0 forcing continuous SYSREF mode
link 1
DEBUG: hmc7044_jesd204_link_pre_setup:1084 link_num 0 reason initialization
DEBUG: hmc7044_jesd204_link_pre_setup:1100 Found SYSREF channel2 setting f=1953125 Hz
DEBUG: hmc7044_jesd204_link_pre_setup:1100 Found SYSREF channel10 setting f=1953125 Hz
WARNING: hmc7044_jesd204_link_pre_setup: Link0 forcing continuous SYSREF mode
op 4
link 0
link 1
op 5
link 0
link 1
op 6
link 0
link 1
op 7
link 0
link 1
op 8
link 0
DEBUG: ad9081_jesd204_setup_stage1:421 reason initialization
link 1
op 9
link 0
link 1
op 10
link 0
link 1
op 11
link 0
link 1
op 12
link 0
link 1
op 13
link 0
DEBUG: ad9081_jesd204_clks_enable:311 reason initialization
link 1
DEBUG: ad9081_jesd204_clks_enable:311 reason initialization
op 14
link 0
DEBUG: ad9081_jesd204_link_enable:363 link_num 0 reason initialization
link 1
DEBUG: ad9081_jesd204_link_enable:363 link_num 0 reason initialization
op 15
link 0
DEBUG: ad9081_jesd204_link_running:385 reason initialization
rx_jesd: Lane 0 desynced (5 errors), restarting link
rx_jesd: Lane 1 desynced (5 errors), restarting link
rx_jesd: Lane 2 desynced (12 errors), restarting link
rx_jesd: Lane 3 desynced (14 errors), restarting link
tx_jesd status:
	Link is enabled
	Measured Link Clock: 250.023 MHz
	Reported Link Clock: 250.000 MHz
	Lane rate: 10000.000 MHz
	Lane rate / 40: 250.000 MHz
	LMFC rate: 7.812 MHz
	SYNC~: deasserted
	Link status: DATA
	SYSREF captured: Yes
	SYSREF alignment error: No
rx_jesd status:
	Link is enabled
	Measured Link Clock: 250.021 MHz
	Reported Link Clock: 250.000 MHz
	Lane rate: 10000.000 MHz
	Lane rate / 40: 250.000 MHz
	LMFC rate: 7.812 MHz
	Link status: DATA
	SYSREF captured: Yes
	SYSREF alignment error: No
tx_dac: Successfully initialized (250022888 Hz)
rx_adc: Successfully initialized (250022888 Hz)
```

### CODE REVIEW

File: drivers/jesd204/jesd204.c [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+
+int jesd204_dev_register(struct jesd204_dev **jdev, char *name, const struct jesd204_dev_data *data)
+{
+	struct jesd204_dev *pjdev;
+	static int id = 0;
+
+	if (jdev == NULL)
+		return -EINVAL;
+
+	pjdev = calloc(1, sizeof(struct jesd204_dev));
```
Comment (cristiantpop) [inline drivers/jesd204/jesd204.c:43]: Use sizeof(*pjdev)

File: drivers/jesd204/jesd204.c [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+
+	topjdev->links = links;
+	topjdev->links[topjdev->num_links] = link_dev;
+	topjdev->num_links += 1;
+
+	if (link_dev->top_device)
+		link_dev->top_device->is_top = true;
+
+	for (d = 0; d < link_dev->num_devices; d++) {
+		struct jesd204_dev *jdev = link_dev->devices[d];
```
Comment (cristiantpop) [inline drivers/jesd204/jesd204.c:75]: Declare variables at the begging of the function

File: drivers/jesd204/jesd204.c [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+}
+
+int jesd204_link_register(struct jesd204_dev *topjdev, struct jesd204_link_dev *link_dev)
+{
+	unsigned int d;
+
+	if (!topjdev || !link_dev)
+		return -EINVAL;
+
+	struct jesd204_link_dev **links = realloc(topjdev->links, sizeof(*topjdev->links) * (topjdev->num_links + 1));
```
Comment (cristiantpop) [inline drivers/jesd204/jesd204.c:63]: Maybe consider list usage:
https://github.com/analogdevicesinc/no-OS/blob/master/include/list.h

File: drivers/jesd204/jesd204.c [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+
+		jdev->links = dlinks;
+		jdev->links[jdev->num_links] = link_dev;
+		jdev->num_links += 1;
+	}
+	
+	return 0;
+}
+
+
```
Comment (cristiantpop) [inline drivers/jesd204/jesd204.c:88]: Extra line

**Comment by buha** [comment]
closing this, jesd204 framework will be added in a separate PR by @dbogdan

**Pull request closed without merging**